### ISSUE TITLE
Instantiate UI Instance with Instantiator class

### ISSUE THREAD

**Issue by mkn1988** [vaadin/flow#7446]
When creating the ``UI`` intance in the ``BootstrapHandler`` via this method: `` com.vaadin.flow.server.BootstrapHandler.createAndInitUI(Class<? extends UI>, VaadinRequest, VaadinResponse, VaadinSession)`` the ``ReflectTools`` are used.

It would be handy, if instead the ``Instantiator`` would be used. Then we would be able to manage the ``UI`` instance via Spring. The ``Instantiator`` is easy to get in the request context via ``request.getService().getInstantiator()``. The ``DefaultInstantiator`` does the same call as the current implementation and the ``SpringInstantiator`` does some "magic", such that we are able to use dependency injection in our custom ``UI`` implementation.

**Flow Version 2.1.4 (but also present in current 2.2.0-alpha12)**

A related issue to this is #6395

**Comment by jhult** [comment]
This seemed pretty easy to implement. You can see my implementation in pull request #7650.

**Comment by Legioth** [comment]
Creating UI instances in his way won't help with #6395 since the information needed for choosing whether to preserve or not isn't available when the UI instance needs to be created.

In general, we want to help developers use a custom UI class less rather than more. This kind of change would go against that idea.

**Comment by jhult** [comment]
I agree that it won't help with #6395.

But, it does seem to address this issue. Whether or not Vaadin wants to generally promote this, is up to y'all. Personally, I don't see the harm in having it available if a developer wants to use it.

**Comment by Legioth** [comment]
From my point of view, there should never be any reason to have a subclass of UI with custom logic, and it would thus also not matter how that instance is created. Apparently, I'm mistaken and we would therefore need to understand what that reason is. Without this, we cannot determine the relevance of the https://github.com/vaadin/flow/pull/7650 PR or alternatively suggest some other way of achieving the same goal.

In other words: what is the use case for this feature suggestion?

**Comment by knoobie** [comment]
Subclassing of UI could be a relic of the past (V8), where a lot of developer added custom code in the UI class. 

For V14+ the only use-case I'm aware of, could be the MockedUI implementation from @mvysny.

**Comment by alvegh** [comment]
In our case, we subclassed UI so we can store a reference to a Router instance inside. 
This is because we have page-scope, where available views  can be different for each page.

**Comment by thigg** [comment]
I have a usecase where sessions were handled by tab, thus the UI instance was used to store all data which was needed to handle the state of the specific tab (not the session, thus the UI instance)

**Comment by denis-anisimov** [comment]
>In our case, we subclassed UI so we can store a reference to a Router instance inside.
This is because we have page-scope, where available views can be different for each page.

`UI::getRouter` allows to get the router.

**Comment by denis-anisimov** [comment]
>I have a usecase where sessions were handled by tab, thus the UI instance was used to store all data which was needed to handle the state of the specific tab (not the session, thus the UI instance)

`ComponentUtil::setData` methods allows to store any data inside any component including `UI`

**Comment by thigg** [comment]
> have a usecase where sessions were handled by tab, thus the UI instance was used to store all data which was needed to handle the state of the specific tab (not the session, thus the UI instance)

If I understand you correctly, I would need a component which always statys the same in the tab, to be able to store data in that tab. While that actually fits my usecase (I store the data now in the base RouterLayout), it feels a bit strange to use a component for that.

**Comment by denis-anisimov** [comment]
If my stat object is represented a `State` class then to be able to store it in UI I may do 2 things:
- extends `UI` class and make a field with `State` type to store the state.
- Use `ComponentUtil.setData(ui, state)` / `ComponentUtil.getData(ui, State.class)`.

Both do the same thing.
As IU understood you have chosen the first option.
Use the second option.

**Comment by thigg** [comment]
@denis-anisimov that works even if the ui is reinstantiated?
When I tried it, it returned null, everytime I refreshed the page.

**Comment by denis-anisimov** [comment]
The same happens if you store data in `UI` subclass.
`UI` is 1:1 to browser tab.

I'm not sure anymore which issue we are talking about here?
Everything which you want to store inside `UI` you may store using via `ComponentUtil::setData`: there is no need to extend `UI`  class.

**Comment by thigg** [comment]
As I wrote, that maybe belongs more to #6395 . I am currently searching for a place where I can store data that corresponds to a tab. In vaadin 8 you could use the UI instance for that.
You could refresh the tab and do whats-o-ever and still access the same data.

@denis-anisimov I hope that clarified it a bit.

**Comment by denis-anisimov** [comment]
Unfortunately I didn't get the clarification....
The ticket title is "Instantiate UI Instance with Instantiator class".
Your original  comment was `thus the UI instance was used to store all data` .
So as I understood you extended `UI` class to be able to store some data.
I've explained how to store data in `UI` class without extending this: so there is no need 
to extend `UI` and no need to instantiate UI with `Instantiator` class (which the ticket is about).

But now I see some sentences about `UI` reinstantiation .
I don't see how `Instantiator` help here.

Ticket  #6395 is about totally different thing and is out of the scope of this issue.

**Comment by thigg** [comment]
@denis-anisimov Sorry, I understood this as a way to circumvent the issues with #6395 or at least solve the usecase when you need to store data per tab and not per session.

**Comment by pleku** [comment]
Hello. I've closed (not merged) the PR that would have done what the title of this issue says, _using the Instantiator to instantiate the UI_. We agreed that we don't want to do that change now. This issue seems like a bit of requesting for things described in #6395 and maybe some other things that "would be beneficial for using our custom UI". But since v10+ the idea is that you should not need to use a custom UI class in your app (other than maybe when migrating from v7/v8 with MPR) and any case where you desire to do so, should be supported by other means than creating a custom UI class. So as it is not clear in scope of this issue that _what is the job that the custom UI class is hired to do_, I'm closing this issue as wontfix.

Please create a new issue and describe what you want to achieve and why, so it may get a proper fix for it. Or +1 #6395 if that is what you really need.
